File: DevExpress VCL/ExpressBars 7/Demos/Delphi/MDIDemo/MDIDemoChild.pas
unit MDIDemoChild;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls, ComCtrls;

type
  TfrmMDIChild = class(TForm)
    Editor: TRichEdit;
    procedure FormCreate(Sender: TObject);
    procedure FormClose(Sender: TObject; var Action: TCloseAction);
    procedure FormActivate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure FormCloseQuery(Sender: TObject; var CanClose: Boolean);
    procedure EditorMouseDown(Sender: TObject; Button: TMouseButton;
      Shift: TShiftState; X, Y: Integer);
  private
    FFileName: string;
    procedure SetFileName(Value: string);
  public
    property FileName: string read FFileName write SetFileName;
  end;

implementation

uses
  FileCtrl, MDIDemoMain;

{$R *.DFM}

procedure TfrmMDIChild.SetFileName(Value: string);
var
  I: Integer;
begin
  FFileName := Value;
  Caption := MinimizeName(FFileName, Canvas, 100);
  with frmMDIMain.dxBarListWindows do
  begin
    I := Items.IndexOfObject(Self);
    if (0 <= I) and (I < Items.Count) then Items[I] := FFileName;
  end;
end;

procedure TfrmMDIChild.FormCreate(Sender: TObject);
begin
  Inc(frmMDIMain.CreatedMDICount);
  Caption := 'Document' + IntToStr(frmMDIMain.CreatedMDICount);
  Editor.OnChange := frmMDIMain.EditorChange;
  Editor.OnSelectionChange := frmMDIMain.EditorSelectionChange;
  frmMDIMain.dxBarListWindows.Items.AddObject(Caption, Self);
  Editor.Modified;

  if frmMDIMain.MDIChildCount = 1 then frmMDIMain.ShowItems(True);
end;

procedure TfrmMDIChild.FormClose(Sender: TObject; var Action: TCloseAction);
begin
  if frmMDIMain.FileName <> '' then
    frmMDIMain.dxBarMRUFiles.AddItem(frmMDIMain.FileName, nil);
  Action := caFree;
end;

procedure TfrmMDIChild.FormActivate(Sender: TObject);
begin
  Editor.OnChange(Editor);
end;

procedure TfrmMDIChild.FormDestroy(Sender: TObject);
begin
  with frmMDIMain.dxBarListWindows.Items do
    Delete(IndexOfObject(Self));

  if frmMDIMain.MDIChildCount = 1 then frmMDIMain.ShowItems(False);
end;

procedure TfrmMDIChild.FormCloseQuery(Sender: TObject;
  var CanClose: Boolean);
var
  Res: Integer;
begin
  if Editor.Modified then
  begin
    Res := Application.MessageBox(
      PChar(Format('Do you want to save the changes you made to "%s"?', [Caption])),
      PChar(Application.Title), MB_ICONQUESTION or MB_YESNOCANCEL);
    case Res of
      ID_YES: CanClose := frmMDIMain.SaveFile(False);
      ID_NO: CanClose := True;
      ID_CANCEL: CanClose := False;
    end;
  end;
end;

procedure TfrmMDIChild.EditorMouseDown(Sender: TObject; Button: TMouseButton;
  Shift: TShiftState; X, Y: Integer);
begin
  if Button = mbRight then
    frmMDIMain.dxBarPopupMenu.PopupFromCursorPos;
end;

end.

File: DevExpress VCL/ExpressBars 7/Demos/CBuilder/MDIDemo/MDIDemoChild.cpp
//---------------------------------------------------------------------------
#include <vcl.h>
#pragma hdrstop

#include "MDIDemoChild.h"
#include <stdio.h>
#include "MDIDemoMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TfrmMDIChild *frmMDIChild;
//---------------------------------------------------------------------------
__fastcall TfrmMDIChild::TfrmMDIChild(TComponent* Owner)
        : TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TfrmMDIChild::SetFileName(AnsiString Value)
{
  int I;

  FFileName = Value;
  Caption = FFileName;
  I = frmMDIMain->dxBarListWindows->Items->IndexOfObject(this);
  if ((0 <= I) && (I < frmMDIMain->dxBarListWindows->Items->Count))
    frmMDIMain->dxBarListWindows->Items->Strings[I] = FFileName;
}
//---------------------------------------------------------------------------

void __fastcall TfrmMDIChild::FormCreate(TObject *Sender)
{
  frmMDIMain->CreatedMDICount++;
  Caption = "Document" + IntToStr(frmMDIMain->CreatedMDICount);
  Editor->OnChange = frmMDIMain->EditorChange;
  Editor->OnSelectionChange = frmMDIMain->EditorSelectionChange;
  frmMDIMain->dxBarListWindows->Items->AddObject(Caption, this);
  Editor->Modified;

  if (frmMDIMain->MDIChildCount == 1) frmMDIMain->ShowItems(True);
}
//---------------------------------------------------------------------------

void __fastcall TfrmMDIChild::FormClose(TObject *Sender,
      TCloseAction &Action)
{
  if (frmMDIMain->FileName != "")
    frmMDIMain->dxBarMRUFiles->AddItem(frmMDIMain->FileName, NULL);

  Action = caFree;
}
//---------------------------------------------------------------------------

void __fastcall TfrmMDIChild::FormActivate(TObject *Sender)
{
  Editor->OnChange(Editor);
}
//---------------------------------------------------------------------------

void __fastcall TfrmMDIChild::FormDestroy(TObject *Sender)
{
  frmMDIMain->dxBarListWindows->Items->Delete(frmMDIMain->dxBarListWindows->Items->IndexOfObject(this));

  if (frmMDIMain->MDIChildCount == 1) frmMDIMain->ShowItems(False);
}
//---------------------------------------------------------------------------

void __fastcall TfrmMDIChild::FormCloseQuery(TObject *Sender, bool &CanClose)
{
  if (Editor->Modified)
  {
	int Res = MessageBox(0,
	   dxStringToAnsiString(Format("Do you want to save the changes you made to ""%s""?", ARRAYOFCONST((Caption))), 0).c_str(),
	   dxStringToAnsiString(Application->Title, 0).c_str(), MB_ICONQUESTION | MB_YESNOCANCEL);
	switch (Res)
    {
      case ID_YES: CanClose = frmMDIMain->SaveFile(False); break;
      case ID_NO: CanClose = True; break;
      case ID_CANCEL: CanClose = False; break;
    }
  }
}
//---------------------------------------------------------------------------

void __fastcall TfrmMDIChild::EditorMouseDown(TObject *Sender,
      TMouseButton Button, TShiftState Shift, int X, int Y)
{
  if (Button == mbRight)
	frmMDIMain->dxBarPopupMenu->PopupFromCursorPos();
}
//---------------------------------------------------------------------------

